### ISSUE TITLE
workbox-precache caches error responses

### ISSUE THREAD

**Issue by bukharin** [GoogleChrome/workbox#2737]
Hi

**Library Affected**:
workbox-precache

**Browser & Platform**:
Google Chrome latest

**Issue or Feature Request Description**:

Workbox precache module caches 404 error response, but it shouldn't, See screenshot:

![image](https://user-images.githubusercontent.com/3126568/105978768-bdd27c00-60b4-11eb-8d3d-7ccd61087f02.png)

Using workbox 6.0.2

My sw.js

```
const precacheUrls = self.__WB_MANIFEST;
if (precacheUrls) {
  precacheAndRoute(precacheUrls);
}
```

**Comment by jeffposnick** [comment]
Hello!

The expected behavior is that any responses with 4xx or 5xx statuses should not be precached by default, and cause the service worker installation to fail:

https://github.com/GoogleChrome/workbox/blob/3151c3a2740558b87a592f5091cc795815ca3ac5/packages/workbox-precaching/src/PrecacheStrategy.ts#L148-L154

Are you possibly passing in a `cacheWillUpdate` plugin when you set up precaching? That's the one situation in which you might end up with a `404` response cached. I know that you included a snippet of your service worker code above, but is there any chance that there's more code you didn't include?

**Comment by jeffposnick** [comment]
...what also leads me to believe that your screenshot either shows a cache created by a much earlier version of Workbox than v6.0.2, or alternatively, that your service worker file includes additional code that you didn't include, is that your precache is named `talk-precache-v1`.

The default precache name in Workbox v6.0.2 ends in `precache-v2`:

https://github.com/GoogleChrome/workbox/blob/3151c3a2740558b87a592f5091cc795815ca3ac5/packages/workbox-core/src/_private/cacheNames.ts#L30-L36

**Comment by bukharin** [comment]
@jeffposnick our project start using Workbox from v5, cache names was not changed. Later project have been upgraded to v6. 

This is full sw.js:

```
const appName = 'talk';

setCacheNameDetails({
  prefix: appName,
  suffix: 'v1',
  precache: 'precache',
  runtime: 'runtime-cache'
});

self.skipWaiting();
self.addEventListener('activate', () => self.clients.claim());

// Cache page navigations (html) with a Network First strategy

registerRoute(
  new NavigationRoute(// Use a Network First caching strategy
    new NetworkFirst({
      plugins: [
        // Ensure that only requests that result in a 200 status are cached
        new CacheableResponsePlugin({
          statuses: [200],
        }),
      ],
    })
  )
);

const precacheUrls = self.__WB_MANIFEST;
if (precacheUrls) {
  precacheAndRoute(precacheUrls);
}

// other routes
```

**Comment by jeffposnick** [comment]
Hmm... we do have a test case for this behavior:

https://github.com/GoogleChrome/workbox/blob/3151c3a2740558b87a592f5091cc795815ca3ac5/test/workbox-precaching/sw/test-PrecacheStrategy.mjs#L105-L117

I'm surprised by what you're seeing.

Is there perhaps a 3xx redirect involved here? In other words, you include a URL that doesn't exist in your precache manifest, and then the server responds first with a 3xx response pointing to another URL, and then that other URL results in the 404?

Is there any chance you could DM `@jeffposnick` on Twitter the link to a live version of your site exhibiting this behavior, if you can't share the URL publicly?

**Comment by bukharin** [comment]
@jeffposnick I can share URL, but the problem not reproduce there. I catched this situation when we deploying a new version and some of our server nodes not received specified `js` files and return 404 response code.

But I can easy reproduce the problem in test branch by adding unexisting js file path:

``` javascript
const precacheUrls = self.__WB_MANIFEST;
if (precacheUrls) {
  precacheAndRoute([
    ...precacheUrls,
    {"revision": null, "url": "not-found.js"}
  ]);
}
```

First installation of service-worker fails, but in Cache Storage I could find `not-found.js` record with 404 response

![image](https://user-images.githubusercontent.com/3126568/106112081-a5bf3300-616e-11eb-84ee-cb5f66591269.png)

Service worker was not successfully installed:

![image](https://user-images.githubusercontent.com/3126568/106112186-c12a3e00-616e-11eb-96fb-1079ea11ca33.png)

But if I refresh the page - the service-worker will be installed successfully and continue to work:

![image](https://user-images.githubusercontent.com/3126568/106112316-edde5580-616e-11eb-8599-e1355f5391be.png)

After that SW will response from Cache Storage:

![image](https://user-images.githubusercontent.com/3126568/106112438-12d2c880-616f-11eb-9190-7f43e7a9f83e.png)

Is it enough information to reproduce the problem? If not I can try to create separate app to illustrate this behaviour.

**Comment by bukharin** [comment]
How we build SW:

1. We have source `sw.js` in ES6 format
2. Build and minify it using Rollup -  `rollup -c rollup-sw.config.js`
3. Inject workbox manifest, using workbox-cli `workbox injectManifest`

`rollup-sw.config.js`:

```
import resolve from "rollup-plugin-node-resolve";
import { terser } from "rollup-plugin-terser";
import replace from "rollup-plugin-replace";

export default {
    input: `src/app/service-worker/sw.js`,
    output: [
        {file: 'src/app/service-worker/sw.bundle.js', format: "esm"}
    ],
    plugins: [
        resolve(),
        replace({
            'process.env.NODE_ENV': JSON.stringify('production'),
        }),
        terser()
    ]
};
```

workbox-config.js`:

```
module.exports = {
  "globDirectory": "dist/",
  "globPatterns": ["**/*.{css,html,js,json,ttf,woff,woff2}"],
  "globFollow": true,
  "globStrict": true,
  "globIgnores": [
    '**/*-es5.*.js',
    'sw.bundle.js',
    'index.html',
  ],
  "dontCacheBustURLsMatching": new RegExp('.+.[a-f0-9]{20}..+'),
  "maximumFileSizeToCacheInBytes": 5000000,
  "swSrc": "src/app/service-worker/sw.bundle.js",
  "swDest": "dist/sw.bundle.js"
};

```

**Comment by bukharin** [comment]
Also I tried to swap steps of bundling and injecting (I think it's more right way), but it has no effect

**Comment by jeffposnick** [comment]
Ah, thank you for sharing all those steps. I believe this is a regression due to some changes in our precaching logic introduced in Workbox v6:

https://github.com/GoogleChrome/workbox/blob/590d1264f57392b1feb538440858f20a52fe6fdf/packages/workbox-precaching/src/PrecacheStrategy.ts#L142-L166

That `const response = await handler.fetchAndCachePut(request)` line should just be a `handler.fetch()`, and the `handler.cachePut()` should be deferred until after the `responseSafeToPrecache` result is confirmed to be `true`. In the current implementation, we're ending up saving entries to the cache and failing the current `install`, but the next `install` attempt succeeds (because we don't attempt to re-cache entries that are already cached unless their `revision` has changed).

I'll get a fix out for this targeted for the upcoming `6.1.0` release. Thanks for pointing it out!

CC: @philipwalton

**Comment by jeffposnick** [comment]
This is now fixed in the v6.1.0 release.

**Comment by huykon** [comment]
I am using version v6.2.4 and the error still there
